packages/solana/src/client.ts

```diff
@@ -0,0 +1,482 @@
+        }
+
+        const signature = await provider.request({
+          method: 'personal_sign',
+          params: [message, this.getAddress()]
+        })
+
+        return signature as string
+      }
+    }
+
+    super({
+      networkControllerClient,
+      connectionControllerClient,
+      defaultChain: SolHelpersUtil.getChainFromCaip(
+        chains,
+        typeof window === 'object' ? localStorage.getItem(SolConstantsUtil.CAIP_CHAIN_ID) : ''
+      ),
+      tokens: HelpersUtil.getCaipTokens(tokens),
+      _sdkVersion: _sdkVersion ?? `html-solana-${ConstantsUtil.VERSION}`,
+      ...w3mOptions
+    } as ScaffoldOptions)
+
+    this.chains = chains
+    this.syncRequestedNetworks(chains, chainImages)
+    SolStoreUtil.setProjectId(options.projectId)
+    const chain = SolHelpersUtil.getChainFromCaip(
+      chains,
+      typeof window === 'object' ? localStorage.getItem(SolConstantsUtil.CAIP_CHAIN_ID) : ''
+    )
+    if (chain) {
+      SolStoreUtil.setCurrentChain(chain)
+      SolStoreUtil.setCaipChainId(`${chain.name}:${chain.chainId}`)
+    }
+    this.syncNetwork(chainImages)
+
+    this.walletAdapters = createWalletAdapters()
+    this.WalletConnectConnector = new WalletConnectConnector({
+      relayerRegion: 'wss://relay.walletconnect.com',
+      metadata: {
+        description: 'Solana in Wallet Connect',
+        name: 'Wallet Connect',
+        icons: ['https://avatars.githubusercontent.com/u/37784886'],
+        url: 'https://web3modal.com'
+      },
+      autoconnect: false,
+      qrcode: true
+    })
+    SolStoreUtil.setConnection(
+      new Connection(chain?.rpcUrl ?? 'https://api.devnet.solana.com', 'recent')
+    )
+
+    SolStoreUtil.subscribeKey('address', () => {
+      this.syncAccount()
+    })
+
+    SolStoreUtil.subscribeKey('caipChainId', () => {
+      this.syncNetwork(chainImages)
+    })
+
+    if (typeof window === 'object') {
+      setTimeout(() => {
+        this.checkActiveProviders()
+        this.syncConnectors()
+      }, 500)
+    }
+  }
+
+  public setAddress(address?: string) {
+    SolStoreUtil.setAddress(address ?? '')
+  }
+
+  public disconnect() {
+    const provider = SolStoreUtil.state.provider as Provider
+
+    if (provider) {
+      provider.emit('disconnect')
+    }
+  }
+
+  public getAddress() {
+    const { address } = SolStoreUtil.state
+
+    return address ? SolStoreUtil.state.address : address
+  }
+
+  public async checkActiveProviders() {
+    const walletId = localStorage.getItem(SolConstantsUtil.WALLET_ID)
+
+    try {
+      if (walletId === ConstantsUtil.WALLET_CONNECT_CONNECTOR_ID) {
+        await this.WalletConnectConnector.connect(true)
+        const provider = await this.WalletConnectConnector.getProvider()
+        const accounts = await provider.enable()
+        this.setWalletConnectProvider(accounts[0])
+      } else {
+        const wallet = walletId?.split('_')[1] as AdapterKey
```

Comment:
do walletIds ALWAYS have `_` ? I'd add optional chaining just in case to prevent this to crash in case it doesn't

Reply:
yeah, all injected wallets have name in id (like injected_phantom, injected_solflare) to distinguish them